src/main/java/hudson/plugins/s3/cloudfront/CloudFrontInvalidatePublisher.java

```diff
@@ -0,0 +1,188 @@
+		logger.println(StringUtils.defaultString(getDescriptor().getDisplayName()) + " " + message);
+	}
+
+	@Override
+	public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+
+		final boolean buildFailed = build.getResult() == Result.FAILURE;
+
+		S3Profile profile = getProfile();
+		if (profile == null) {
+			log(listener.getLogger(), "No S3 profile is configured.");
+			build.setResult(Result.UNSTABLE);
+			return true;
+		}
+		log(listener.getLogger(), "Using S3 profile: " + profile.getName());
+		try {
+			Map<String, String> envVars = build.getEnvironment(listener);
+			for (InvalidationEntry entry : invalidationEntries) {
+
+				if (entry.noInvalidateOnFailure && buildFailed) {
+					// build failed. don't post
+					log(listener.getLogger(), "Skipping S3 key invalidation because build failed");
+					continue;
+				}
+
+				String keyPrefix = Util.replaceMacro(entry.keyPrefix, envVars);
+				if (StringUtils.isBlank(keyPrefix)) {
+					log(listener.getLogger(), "No S3 asset key was provided.");
+					continue;
+				}
+
+				String bucket = Util.replaceMacro(entry.bucket, envVars);
+				if (StringUtils.isBlank(bucket)) {
+					log(listener.getLogger(), "S3 bucket was not specified.");
+					continue;
+				}
+				
+				InvalidationRecord invalidationRecord = profile.invalidate(build, listener, bucket, keyPrefix);
+				log(listener.getLogger(), "With bucket = " + bucket 
+						+ "; keyPrefix = " + keyPrefix + ";\n\t" + invalidationRecord);
+			}
+		} catch (IOException e) {
+			e.printStackTrace(listener.error("Failed to upload files"));
+			build.setResult(Result.UNSTABLE);
+		}
+		return true;
+	}
+
+	@Override
+	public BuildStepDescriptor<Publisher> getDescriptor() {
+		return DESCRIPTOR;
+	}
+	
+	@Override
+	public BuildStepMonitor getRequiredMonitorService() {
+		return BuildStepMonitor.STEP;
+	}
+
+	public static final class DescriptorImpl extends BuildStepDescriptor<Publisher> {
+
+		private final CopyOnWriteList<S3Profile> profiles = new CopyOnWriteList<S3Profile>();
+
+		public DescriptorImpl(Class<? extends Publisher> clazz) {
+			super(clazz);
+			load();
+		}
+
+		public DescriptorImpl() {
+			this(CloudFrontInvalidatePublisher.class);
+		}
+
+		@Override
+		public String getDisplayName() {
+			return "Invalidate S3 assets among CloudFront distributions";
+		}
+
+		@Override
+		public String getHelpFile() {
+			return "/plugin/s3/help-invalidate.html";
+		}
+
+		public ListBoxModel doFillProfileNameItems() {
+			ListBoxModel model = new ListBoxModel();
+			
+			for (S3Profile profile : getProfiles()) {
+				model.add(profile.getName(), profile.getName());
+			}
+			return model;
+		}
+		
+		public S3Profile[] getProfiles() {
+			if(profiles.isEmpty()){
+				profiles.addAll(Arrays.asList(((hudson.plugins.s3.S3BucketPublisher.DescriptorImpl) Jenkins.getInstance().getDescriptor(S3BucketPublisher.class)).getProfiles()));
```

Comment:
this line is too long, could you please extract some parts like getting descriptors?